File: utilities/frir_generator.m
function [h, beta_hat] = frir_generator(c, fs, r, s, L, beta, nSamples)
% FRIR_GENERATOR Generates a room impulse response from the given parameters.
% c = speed of sound (m/s)
% fs = sampling frequency of impulse response
% r = 1 x 3 array specifying the (x,y,z) coordinates of receiver (m)
% s = 1 x 3 array specifying the (x,y,z) coordinates of source (m)
% L = 1 x 3 array specifying the (x,y,z) room dimensions (m)
% beta = T60 reverberation time (s)
% nSamples = total number of samples in impulse response
% h = nsample x 1 column vector containing the calculated room impulse response
% beta_hat = reflection coefficient of the walls based on reverberation time
%
% Current algorithm: based on the fast image method by McGovern [1].
%
% [1] Stephen G. McGovern, Fast image method for impulse response calculations
% of box-shaped rooms, Applied Acoustics 70 (2009), 182-189.
%
    % Require all arguments
    if nargin < 7, error('Not enough input arguments.'); end

    % Calculate reflection coefficient via Sabine's formula
    if beta > 0
        volume = L(1) * L(2) * L(3);
        surfaceArea = 2 * (L(1)*L(3) + L(2)*L(3) + L(1)*L(2));
        alpha = 24 * volume * log(10.0) / (c * surfaceArea * beta);
        if alpha > 1
            error(['Error: Reflection coefficient cannot be determined ' ...
                'using the current room parameters. Please change the room ' ...
                'size or reverberation time.']);
        end
        beta_hat = sqrt(1 - alpha);
    else
        error('Error: Reverberation time must be positive.\n');
    end

    % Calculate total time of impulse response in seconds
    time = nSamples / fs;

    % Find highest-order reflection to be used in lookup tables
    xMaxOrder = ceil(c * time / (2 * L(1)));
    yMaxOrder = ceil(c * time / (2 * L(2)));
    zMaxOrder = ceil(c * time / (2 * L(3)));

    % Create lookup tables for each dimension
    [xDistSq, xRefCoef] = make_tables( ...
        L(1), r(1), s(1), beta_hat, beta_hat, xMaxOrder);
    [yDistSq, yRefCoef] = make_tables( ...
        L(2), r(2), s(2), beta_hat, beta_hat, yMaxOrder);
    [zDistSq, zRefCoef] = make_tables( ...
        L(3), r(3), s(3), beta_hat, beta_hat, zMaxOrder);

    % Initialize impulse response array
    h = zeros(nSamples, 1);

    i = 1;
    while i <= 2*xMaxOrder + 1
        j = 1;
        while j <= 2*yMaxOrder + 1
            k = 1;
            while k <= 2*zMaxOrder + 1
                % Find distance from ijk^th image to receiver
                dist = sqrt(xDistSq(i) + yDistSq(j) + zDistSq(k));

                % Find sample in impulse response corresponding to distance
                samp = round(dist * fs / c) + 1;

                % If echo from ijk^th image occurs at a time beyond range of
                % impulse response, then all images in one strip of circle have
                % been tested.
                if samp > nSamples, break; end

                % Add echo to impulse response array
                h(samp) = h(samp) + ...
                    (xRefCoef(i) * yRefCoef(j) * zRefCoef(k) / dist);

                k = k + 1;
            end

            % If k-loop broke with k == 1, then all images in "circle of
            % interest" in i-plane have been tested, so no need to find
            % distances for higher values of j.
            if k == 1, break; end
            j = j + 1;
        end

        % If j-loop broke with j == 1, then all images in "sphere of interest"
        % have been tested, so no need to find distances for higher values of i.
        if j == 1, break; end
        i = i + 1;
    end
    end

    % Create square-of-distance and reflection coefficient lookup tables
    function [distSq, refCoef] = make_tables(w, r, s, Bmw, Bw, n)
    Limg = 1;
    Rimg = 1;

    % Initialize tables
    distSq  = zeros(2*n + 1, 1);
    refCoef = zeros(2*n + 1, 1);

    % Set initial values of tables
    refCoef(1) = 1;
    DirSnd = s - r;
    distSq(1) = DirSnd * DirSnd;

    % Set initial values of left/right image distance sequences
    LimgD = DirSnd;
    RimgD = DirSnd;

    % Create tables
    for i = 2:(n + 1)
        % Set image distances and reflection coefficients for i^th order images
        % on left and right
        if mod(i, 2) ~= 0
            LimgD = LimgD - (2 * (w + s));
            RimgD = RimgD + (2 * (w - s));
            Limg = Limg * Bmw;
            Rimg = Rimg * Bw;
        else
            LimgD = LimgD - (2 * (w - s));
            RimgD = RimgD + (2 * (w + s));
            Limg = Limg * Bw;
            Rimg = Rimg * Bmw;
        end

        % Determine pattern for i^th order echo pair and set values for both
        % tables
        if (mod(i, 2) ~= 0 && r + s < 0) || (mod(i, 2) == 0 && r - s < 0)
            distSq(2*i - 2) = LimgD * LimgD;
            distSq(2*i - 1) = RimgD * RimgD;
            refCoef(2*i - 2) = Limg;
            refCoef(2*i - 1) = Rimg;
        else
            distSq(2*i - 2) = RimgD * RimgD;
            distSq(2*i - 1) = LimgD * LimgD;
            refCoef(2*i - 2) = Rimg;
            refCoef(2*i - 1) = Limg;
        end
    end
end

File: references/frir_generator.c
/*
Program     : Fast Room Impulse Response Generator

Description : Calculates the impulse response of a box-shaped room using the
              fast image method [1].
              For test purposes only.

              [1] Stephen G. McGovern,
              Fast image method for impulse response calculations of box-shaped
              rooms, Applied Acoustics 70 (2009), 182-189.

Author      : Edward Ly (m5222120@u-aizu.ac.jp)
Version     : 0.1.0
*/

#include <stdio.h>
#include <math.h>
#include <stdlib.h>

// Data structure to hold data for a particular dimension
struct axisVar {
  double w;        // room width
  double r;        // distance from receiver to origin
  double s;        // distance from sound source to origin
  double Bmw;      // reflection coefficient of wall to left of origin
  double Bw;       // reflection coefficient of wall to right of origin
  int n;           // highest order reflection for lookup tables
  double* distSq;  // pointer to square-of-distance lookup table
  double* refCoef; // pointer to reflection coefficient lookup table
};

//-----------------------------------------------------------------------------

// Create square-of-distance and reflection coefficient lookup tables
void makeTables (struct axisVar a) {
  double Limg = 1, Rimg = 1, DirSnd;
  double RimgD, LimgD;

  // Set 1st value of reflection coefficient table
  a.refCoef[0] = 1;

  // Set 1st value of square-of-distance table
  DirSnd = a.s - a.r;
  a.distSq[0] = DirSnd * DirSnd;

  // Set initial values of left/right image distance sequences
  LimgD = DirSnd;
  RimgD = DirSnd;

  // Create tables
  for (int i = 1; i <= a.n; i++) {
    // Set image distances and reflection coefficients for i^th order images on
    // left and right
    if (i % 2 != 0) {
      LimgD -= 2 * (a.w + a.s);
      RimgD += 2 * (a.w - a.s);
      Limg  *= a.Bmw;
      Rimg  *= a.Bw;
    } else {
      LimgD -= 2 * (a.w - a.s);
      RimgD += 2 * (a.w + a.s);
      Limg  *= a.Bw;
      Rimg  *= a.Bmw;
    }

    // Determine pattern for i^th order echo pair and set values for both tables
    if ((i % 2 != 0 && a.r + a.s < 0) || (i % 2 == 0 && a.r - a.s < 0)) {
      a.distSq [2*i - 1] = LimgD * LimgD;
      a.distSq [2*i    ] = RimgD * RimgD;
      a.refCoef[2*i - 1] = Limg;
      a.refCoef[2*i    ] = Rimg;
    } else {
      a.distSq [2*i - 1] = RimgD * RimgD;
      a.distSq [2*i    ] = LimgD * LimgD;
      a.refCoef[2*i - 1] = Rimg;
      a.refCoef[2*i    ] = Limg;
    }
  }
}

//-----------------------------------------------------------------------------

// main function
int main () {
  double c, time, *h, dist;
  int length, sampleRate, samp;
  int i, j, k;

  // Declare 3 structs to hold info for each dimension
  struct axisVar x, y, z;

  // Set speed of sound, total time, and sample rate
  c = 343.0;
  time = 1.5;
  sampleRate = 44100;

  // Calculate length of impulse response in number of samples
  length = (int)((double)sampleRate * time);

  // Set room dimensions (width), source position, receiver position
  x.w = 5; y.w = 5; z.w = 4.5;
  x.r = 4; y.r = 2; z.r = 2;
  x.s = 1; y.s = 1; z.s = 0.5;

  // Set reflection coefficients for each wall
  x.Bmw = 0.7; x.Bw = 0.7;
  y.Bmw = 0.7; y.Bw = 0.7;
  z.Bmw = 0.7; z.Bw = 0.7;

  // Find highest-order reflection to be used in lookup tables
  x.n = (int)ceil(c * time / (2 * x.w));
  y.n = (int)ceil(c * time / (2 * y.w));
  z.n = (int)ceil(c * time / (2 * z.w));

  // Initialize lookup tables
  x.distSq  = (double*)malloc((2*x.n + 1) * sizeof(double));
  y.distSq  = (double*)malloc((2*y.n + 1) * sizeof(double));
  z.distSq  = (double*)malloc((2*z.n + 1) * sizeof(double));
  x.refCoef = (double*)malloc((2*x.n + 1) * sizeof(double));
  y.refCoef = (double*)malloc((2*y.n + 1) * sizeof(double));
  z.refCoef = (double*)malloc((2*z.n + 1) * sizeof(double));

  // Create tables for each dimension
  makeTables(x);
  makeTables(y);
  makeTables(z);

  // Initialize impulse response array
  h = (double*)calloc(length, 8);

  // Use lookup tables to calculate impulse response
  for (i = 0; i < 2*x.n + 1; i++) {
    for (j = 0; j < 2*y.n + 1; j++) {
      for (k = 0; k < 2*z.n + 1; k++) {
        // Find distance from ijk^th image to receiver
        dist = sqrt(x.distSq[i] + y.distSq[j] + z.distSq[k]);

        // Find sample in impulse response corresponding to distance
        samp = (int)(dist * (double)(sampleRate) / c + 0.5);

        // If echo from ijk^th image occurs at a time beyond range of impulse
        // response, then all images in one strip of circle have been tested.
        if (samp >= length) break;

        // Add echo to impulse response array
        h[samp] += x.refCoef[i] * y.refCoef[j] * z.refCoef[k] / dist;
      }

      // If k-loop broke with k == 0, then all images in "circle of interest"
      // in i-plane have been tested, so no need to find distances for higher
      // values of j.
      if (k == 0) break;
    }

    // If j-loop broke with j == 0, then all images in "sphere of interest"
    // have been tested, so no need to find distances for higher values of i.
    if (j == 0) break;
  }

  // Free allocated memory
  free(h);
  free(x.distSq);
  free(y.distSq);
  free(z.distSq);
  free(x.refCoef);
  free(y.refCoef);
  free(z.refCoef);

  return 0;
}
